File: BE/src/main/java/com/react/auth/service/JwtService.java
package com.react.auth.service;
import com.react.user.entity.User;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;
import java.security.Key;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

@Service
public class JwtService {

    private final String secretKey;
    private final long accessExpMs;

    public JwtService(
            @Value("${jwt.secret-key}") String secretKey,
            @Value("${app.jwt.access-exp-ms:60000}") long accessExpMs
    ) {
        this.secretKey = secretKey;
        this.accessExpMs = accessExpMs;
    }

    public String extractUserName(String token) {
        return extractClaim(token, Claims::getSubject);
    }

    private Date extractExpireDate(String token) {
        return extractClaim(token, Claims::getExpiration);
    }

    private boolean isTokenExpired(String token) {
        return extractExpireDate(token).before(new Date());
    }

    public boolean isTokenValid(String token, UserDetails userDetails) {
        String email = extractUserName(token);
        return email.equals(userDetails.getUsername()) && !isTokenExpired(token);
    }

    public String generateToken(UserDetails userDetails) {
        Map<String, Object> claims = new HashMap<>();

        if (userDetails instanceof User user) {
            claims.put("role", user.getRole().name());
        }
        return generateToken(claims, userDetails);
    }

    private String generateToken(
            Map<String, Object> extraClaims,
            UserDetails userDetail
    ) {
        long now = System.currentTimeMillis();

        return Jwts.builder()
                .setClaims(extraClaims)
                .setSubject(userDetail.getUsername())
                .setIssuedAt(new Date(now))
                .setExpiration(new Date(now + accessExpMs))
                .signWith(getSignInKey(), SignatureAlgorithm.HS256)
                .compact();
    }

    public <T> T extractClaim(String token, java.util.function.Function<Claims, T> claimsResolver) {
        Claims claims = extractAllClaims(token);
        return claimsResolver.apply(claims);
    }

    // Tách token: Header.Payload.Signature
    public Claims extractAllClaims(String token) {
        return Jwts
                .parserBuilder()
                .setSigningKey(getSignInKey())
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    public Key getSignInKey() {
        // decode base64 secret key → byte[]
        byte[] keyBytes = Decoders.BASE64.decode(secretKey);
        // tạo HMAC-SHA key dùng cho sign & verify
        return Keys.hmacShaKeyFor(keyBytes);
    }
}

File: BaseSpringSecurity/src/main/kotlin/auth/service/JwtService.kt
package com.auth.service

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service
import java.security.Key
import java.util.Date

@Service
class JwtService(
    @Value("\${jwt.secret-key}") private val secretKey: String,
    @Value("\${app.jwt.access-exp-ms:60000}") private val accessExpMs: Long,
) {
    fun extractUserName(token: String): String {
        return extractClaim(token, Claims::getSubject)
    }

    private fun extractExpireDate(token: String): Date {
        return extractClaim(token, Claims::getExpiration)
    }

    private fun isTokenExpired(token: String): Boolean {
        return extractExpireDate(token).before(Date())
    }

    fun isTokenValid(token: String, userDetails: UserDetails): Boolean {
        val email = extractUserName(token)
        return (email.equals(userDetails.username)) && !isTokenExpired(token)
    }

    fun generateToken(userDetails: UserDetails): String {
        return generateToken(emptyMap(), userDetails)
    }

    private fun generateToken(
        extraClaims: Map<String, Any>,
        userDetail: UserDetails
    ): String {
        val now = System.currentTimeMillis()
        return Jwts
            .builder()
            .setClaims(extraClaims)
            //ghi đè key tương ứng trong claims
            .setSubject(userDetail.username)
            .setIssuedAt(Date(System.currentTimeMillis()))
            .setExpiration(Date(now + accessExpMs))
            .signWith(getSignInKey(), SignatureAlgorithm.HS256)
            .compact()
    }

    fun <T> extractClaim(token: String, claimsResolver: (Claims) -> T): T {
        val claims: Claims = extractAllClaims(token)
        return claimsResolver(claims)
    }

    // tách token ra thành 3 phần Header.Payload.Signature
    private fun extractAllClaims(token: String): Claims {
        return Jwts
            .parserBuilder()
            .setSigningKey(getSignInKey())
            .build()
            .parseClaimsJws(token)
            .getBody()
    }

    fun getSignInKey(): Key {
        // lấy chuỗi secretKey trong application để chuyển thành mảng byte
        val keyBytes: ByteArray = Decoders.BASE64.decode(secretKey)
        // trả về 1 SecretKey mới có thể đưa vào .signWith() để ký token hoặc .setSigningKey() để verify token.
        return Keys.hmacShaKeyFor(keyBytes)
    }


}